system/jlib/jmetrics.cpp

```diff
@@ -0,0 +1,219 @@
+       http://www.apache.org/licenses/LICENSE-2.0
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "jmetrics.hpp"
+#include "jmutex.hpp"
+#include "jlog.hpp"
+
+using namespace hpccMetrics;
+
+static Singleton<MetricsReporter> metricsReporter;
+MODULE_INIT(INIT_PRIORITY_STANDARD)
+{
+    return true;
+}
+
+MODULE_EXIT()
+{
+    delete metricsReporter.queryExisting();
+}
+
+
+MetricsReporter &hpccMetrics::queryMetricsReporter()
+{
+    return *metricsReporter.query([] { return new MetricsReporter; });
+}
+
+
+void CounterMetric::collect(MeasurementVector &mv) const
+{
+    mv.emplace_back((std::make_shared<Measurement<uint32_t>>(name, description, ValueType::METRICS_UNSIGNED, count.load())));
+}
+
+
+void GaugeMetric::collect(MeasurementVector &mv) const
+{
+    mv.emplace_back((std::make_shared<Measurement<float>>(name, description, ValueType::METRICS_INTEGER, gaugeValue.load())));
+}
+
+
+MetricsReporter::~MetricsReporter()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        delete sinkIt.second.pSink;
+    }
+}
+
+
+void MetricsReporter::init(IPropertyTree *pMetricsTree)
+{
+    Owned<IPropertyTree> pSinkTree = pMetricsTree->getPropTree("sinks");
+    auto sinkElementsIt = pSinkTree->getElements("sink");
+    initializeSinks(sinkElementsIt);
+}
+
+
+void MetricsReporter::addMetric(const std::shared_ptr<IMetric> &pMetric)
+{
+    std::unique_lock<std::mutex> lock(metricVectorMutex);
+    auto it = metrics.find(pMetric->queryName());
+    if (it == metrics.end())
+    {
+        metrics.insert({pMetric->queryName(), pMetric});
+    }
+    // todo - need to decide how to handle duplicate named metrics
+}
+
+
+void MetricsReporter::removeMetric(const std::string &metricName)
+{
+    std::unique_lock<std::mutex> lock(metricVectorMutex);
+    auto it = metrics.find(metricName);
+    if (it != metrics.end())
+    {
+        metrics.erase(it);
+    }
+}
+
+
+void MetricsReporter::startCollecting()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        sinkIt.second.pSink->startCollection(this);
+    }
+}
+
+
+void MetricsReporter::stopCollecting()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        sinkIt.second.pSink->stopCollection();
+    }
+}
+
+
+std::vector<std::shared_ptr<IMetric>> MetricsReporter::queryMetricsForReport(const std::string &sinkName)
+{
+    std::vector<std::shared_ptr<IMetric>> reportMetrics;
+
+    auto it = sinks.find(sinkName);
+    if (it != sinks.end())
+    {
+        //
+        // Lock the list of metrics while it's in use
+        std::unique_lock<std::mutex> lock(metricVectorMutex);   // no one else can mess with it for a bit
```

Comment:
worth reserving the space in the reportMetrics vector before appending, otherwise this could be O(N^2) for very large numbers of metrics.

Reply:
Here is my thinking. The mutex is protecting access to the map of metrics managed by the reporter. The map contains weak pointers, so it does not maintain a reference count. When a sink needs its list of metrics to report, each metric weak pointer is converted to a shared pointer which is an atomic operation. The shared pointer is placed in a vector and returned to the sink, thus ensuring it has a reference until the sink destroys the vector. During that time if another sink requests its list, the pointer is already locked, so the conversion to shared increments the reference count. 

However, during a report let's say the component destroys the metric by clearing the shared pointer. The metric object will remain alive since the report vector contains a shared pointer to the metric and thus a reference. When the report is completed and the sink destroys the report vector, the ref count goes to 0 and the metric is destroyed (note that the weak pointer remains in the managed list by the reporter). The next time a metric report list is requested, the conversion to a shared pointer will fail and the metric is removed from the list. 

If I am missing something, please let me know.

Reply:
I agree with your overall statement.

My comment was in reference to the statement "A planned optimization is to only build this vector when there is a change".  I assumed that mean you were thinking of preserving the vector from call to call.  If that was done the vector would contain shared_ptrs, which would mean the weak_ptr never returned null.